cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -663,15 +668,34 @@ where
 	/// Checks the execution receipt from the executor peers.
 	fn on_execution_receipt(
 		&self,
-		execution_receipt: &ExecutionReceipt<Block::Hash>,
+		signed_execution_receipt: &SignedExecutionReceipt<Block::Hash>,
 	) -> Result<Action, Self::Error> {
-		// TODO: validate the Proof-of-Election
+		let SignedExecutionReceipt { execution_receipt, signature, signer } =
+			signed_execution_receipt;
 
 		let block_hash = execution_receipt.secondary_hash;
 		let primary_hash =
 			PBlock::Hash::decode(&mut execution_receipt.primary_hash.encode().as_slice())
 				.expect("Hash type must be correct");
 
+		let msg = execution_receipt.hash().encode();
+		if !signer.verify(&msg, signature) {
+			return Err(Self::Error::BadExecutionReceiptSignature)
+		}
+
+		let expected_authority_id = self
+			.primary_chain_client
+			.runtime_api()
+			.authority_id(&BlockId::Hash(primary_hash))?;
+		if *signer != expected_authority_id {
+			// TODO: handle the misbehavior.
```

Comment:
can we block the executor on the libp2p layer itself for misbehaving?

Reply:
I _think_ this error should contribute to node's reputation already, but someone needs to double-check it with source code.

Reply:
Actually, I recall that Jeremiah is not a fan of the peer reputation system in Substrate. Anyway, we'll discuss this before proceeding.